### ISSUE TITLE
Computation of gradient w.r.t. Distribution parameters

### ISSUE THREAD

**Issue by MaxInGaussian** [thu-ml/zhusuan#72]
I have detailedly studied the code under the folders zhusuan/distributions and zhusuan/model. Yet, I am still confused by how the gradients are computed by Tensorflow. Since a number of samples are used to caluated the cost at each optimization Op, I am not sure why this is reversible when taking the gradients w.r.t. Distribution parameters (e.g. mean and log_std in Normal), how this is handled by Tensorflow. It is much appreciated if you can suggest me some materials which I have missed to read.

**Comment by thjashin** [comment]
Hi @MaxInGaussian , thanks for your interests. The point is that you can always differentiate through a random variable by doing some math. Specifically, you write out the loss, which is usually in a form of $E_{p(z)} f(z)$, then change the expectation as an integral $\int p(z)f(z) dz$, and take derivatives of it. The key point here is how to get a good estimate of the gradient with low variance. There are many papers about gradient estimators. Basically for Gaussian random variables, you could use the reparameterization trick (see [the VAE paper](https://arxiv.org/pdf/1312.6114.pdf) ). There are many others, see Section 2 of [this paper](https://arxiv.org/pdf/1611.00712.pdf) for a good review.

**Comment by MaxInGaussian** [comment]
Hi @thjashin , thanks for the reply. I was trying to reproduce the [MCDropout model](https://arxiv.org/pdf/1506.02142.pdf) in Zhusuan, The mean-square-error loss function was directly used as the optimization objective. The confusing part is that, I observed that the dropout rate (the parameter to Bernoulli) was well-adjusted during training if I set it as a trainable variable. This is case where I cannot understand how Tensorflow differentiate the mean-square-error loss w.r.t. to StochasticTensor.

**Comment by thjashin** [comment]
I don't think TF will take gradient through Bernoulli random variable though.
```python
p = tf.Variable(0.5)
a = zs.distributions.Bernoulli(p)
samples = a.sample()
print tf.gradients(samples, p)
# ==> [None]
```
If you want to compute gradient through Bernoulli variables, you may need to use gradient estimators like the [reinforce](http://zhusuan.readthedocs.io/en/latest/api/zhusuan.variational.exclusive_kl.html#zhusuan.variational.exclusive_kl.EvidenceLowerBoundObjective.reinforce). Currently this is implemented for variational inference. See [Bernoulli latent VAE](https://github.com/thu-ml/zhusuan/blob/master/examples/variational_autoencoders/bernoulli_latent_vae.py) in our examples.

**Comment by MaxInGaussian** [comment]
Thanks so much for the suggestion. I will try reinforce to update the dropout rate. For Bernoulli distribution, I guess the changes was stem from the random initializations. Besides, for Normal distribution, I found that it is feasible to obtain the gradient. Should I refer to `tf.random_normal` in this case?
`
mu, ls =tf.Variable(0.), tf.Variable(0.)
a=zs.distributions.Normal(mu, logstd=ls)
s=a.sample(10)
g=tf.gradients(tf.reduce_mean(s), mu)
print(sess.run(g))
# ==> [1.0]
`

**Comment by MaxInGaussian** [comment]
Great thanks for your help and the recommended materials. I have figured out how the derivative is obtained for in variational inference or otherwise.

**Comment by thjashin** [comment]
You are welcome :)
